Version 1.0.0

# Introduction

In this programming assignment we will illustrate a very severe data leakage, that can often be found in competitions, where the pairs of object should be scored, e.g. predict $1$ if two objects belong to the same class and $0$ otherwise. 

The data in this assignment is taken from a real competition, and the funniest thing is that *we will not use training set at all* and achieve almost 100% accuracy score! We will just exploit the leakage.

Now go through the notebook and complete the assignment.

In [1]:
import numpy as np
import pandas as pd 
import scipy.sparse

# Load the data

Let's load the test data. Note, that we don't have any training data here, just test data. Moreover, *we will not even use any features* of test objects. All we need to solve this task is the file with the indices for the pairs, that we need to compare.

Let's load the data with test indices.

In [2]:
test = pd.read_csv('../readonly/data_leakages_data/test_pairs.csv')
test.head(10)

pairId  FirstId  SecondId
0       0     1427      8053
1       1    17044      7681
2       2    19237     20966
3       3     8005     20765
4       4    16837       599
5       5     3657     12504
6       6     2836      7582
7       7     6136      6111
8       8    23295      9817
9       9     6621      7672

For example, we can think that there is a test dataset of images, and each image is assigned a unique `Id` from $0$ to $N-1$ (N -- is the number of images). In the dataframe from above `FirstId` and `SecondId` point to these `Id`'s and define pairs, that we should compare: e.g. do both images in the pair belong to the same class or not. So, for example for the first row: if images with `Id=1427` and `Id=8053` belong to the same class, we should predict $1$, and $0$ otherwise. 

But in our case we don't really care about the images, and how exactly we compare the images (as long as comparator is binary).  

**We suggest you to try to solve the puzzle yourself first.** You need to submit a `.csv` file with columns `pairId` and `Prediction` to the grader. The number of submissions allowed is made pretty huge to let you explore the data without worries. The returned score should be very close to $1$.

**If you do not want to think much** -- scroll down and follow the instructions below.

# EDA and leakage intuition

As we already know, the key to discover data leakages is careful EDA. So let's start our work with some basic data exploration and build an intuition about the leakage.

First, check, how many different `id`s are there: concatenate `FirstId` and `SecondId` and print the number of unique elements. Also print minimum and maximum value for that vector.

In [3]:
# YOUR CODE GOES HERE
print("# of First Id:",test.FirstId.count())
print("# of Second Id:",test.SecondId.count())
print("# of unique First Id:",test.FirstId.nunique())
print("# of unique Second Id:",test.SecondId.nunique())
unique_1=list(test.FirstId.unique())
print(unique_1[1])
unique_2=list(test.SecondId.unique())
print(unique_2[1])
unique_3 = unique_1 + unique_2
print(unique_3[1]," ",unique_3[26326])
unique=set(list(test.FirstId)+list(test.SecondId))
print("sum: ",len(unique_3),", len: ",len(unique)," from: ",min(unique)," to: ",max(unique))

# of First Id: 368550
# of Second Id: 368550
# of unique First Id: 26325
# of unique Second Id: 26310
17044
7681
17044   7681
sum:  52635 , len:  26325  from:  0  to:  26324


and then print how many pairs we need to classify (it is basically the number of rows in the test set)

In [4]:
# YOUR CODE GOES HERE
print(len(test),", ",len(test.FirstId),", ",len(test.SecondId),", ",test.FirstId.count())
print(test.FirstId.sum())
print(test.FirstId.value_counts())

368550 ,  368550 ,  368550 ,  368550
4003780192
0        21
1261     21
1517     21
1658     21
123      21
379      21
635      21
891      21
1147     21
1146     21
1403     21
1659     21
124      21
380      21
636      21
892      21
1402     21
890      21
1148     21
377      21
632      21
888      21
1144     21
1400     21
1656     21
121      21
633      21
634      21
889      21
1145     21
         ..
25194     7
25875     7
25801     7
25079     7
25485     7
24938     7
25545     7
25602     7
24632     7
25934     7
25229     7
25619     7
25289     7
25327     7
24493     7
24295     7
25858     7
24973     7
25033     7
24574     7
24888     7
25678     7
24777     7
24527     7
25363     7
25579     7
25706     7
25003     7
24271     7
24899     7
Name: FirstId, Length: 26325, dtype: int64


Now print, how many distinct pairs it would be possible to create out of all "images" in the dataset?   

In [5]:
# YOUR CODE GOES HERE
print((max(test.FirstId.nunique(),test.SecondId.nunique()))*(max(test.FirstId.nunique(),test.SecondId.nunique())-1)*0.5)
print(test.FirstId.count()/((max(test.FirstId.nunique(),test.SecondId.nunique()))*(max(test.FirstId.nunique(),test.SecondId.nunique())-1)*0.5))

346489650.0
0.00106366813554


So the number of pairs we are given to classify is very very small compared to the total number of pairs. 

To exploit the leak we need to **assume (or prove)**, that the total number of positive pairs is small, compared to the total number of pairs. For example: think about an image dataset with $1000$ classes, $N$ images per class. Then if the task was to tell whether a pair of images belongs to the same class or not, we would have $1000\frac{N(N-1)}{2}$ positive pairs, while total number of pairs was $\frac{1000N(1000N - 1)}{2}$.

Another example: in [Quora competitition](https://www.kaggle.com/c/quora-question-pairs) the task was to classify whether a pair of qustions are duplicates of each other or not. Of course, total number of question pairs is very huge, while number of duplicates (positive pairs) is much much smaller.

Finally, let's get a fraction of pairs of class `1`. We just need to submit a constant prediction "all ones" and check the returned accuracy. Create a dataframe with columns `pairId` and `Prediction`, fill it and export it to `.csv` file. Then submit to grader and examine grader's output. 

In [6]:
# YOUR CODE GOES HERE
onesPred_list  = np.append(np.arange(len(test)).reshape(len(test),1),np.ones(len(test)).reshape(len(test),1),axis=1)
print(onesPred_list)
onesPred_pd = pd.DataFrame(onesPred_list, dtype=int, columns=['pairId', 'Prediction'])
onesPred_pd.to_csv("onesPred_pd.csv",index=False)
onesPred_pd

[[  0.00000000e+00   1.00000000e+00]
 [  1.00000000e+00   1.00000000e+00]
 [  2.00000000e+00   1.00000000e+00]
 ..., 
 [  3.68547000e+05   1.00000000e+00]
 [  3.68548000e+05   1.00000000e+00]
 [  3.68549000e+05   1.00000000e+00]]


pairId  Prediction
0            0           1
1            1           1
2            2           1
3            3           1
4            4           1
5            5           1
6            6           1
7            7           1
8            8           1
9            9           1
10          10           1
11          11           1
12          12           1
13          13           1
14          14           1
15          15           1
16          16           1
17          17           1
18          18           1
19          19           1
20          20           1
21          21           1
22          22           1
23          23           1
24          24           1
25          25           1
26          26           1
27          27           1
28          28           1
29          29           1
...        ...         ...
368520  368520           1
368521  368521           1
368522  368522           1
368523  368523           1
368524  368524           1
368525  368525           1
368526  368526           1
368527  368527           1
368528  368528           1
368529  368529           1
368530  368530           1
368531  368531           1
368532  368532           1
368533  368533           1
368534  368534           1
368535  368535           1
368536  368536           1
368537  368537           1
368538  368538           1
368539  368539           1
368540  368540           1
368541  368541           1
368542  368542           1
368543  368543           1
368544  368544           1
368545  368545           1
368546  368546           1
368547  368547           1
368548  368548           1
368549  368549           1

[368550 rows x 2 columns]

So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

# Building a magic feature

In this section we will build a magic feature, that will solve the problem almost perfectly. The instructions will lead you to the correct solution, but please, try to explain the purpose of the steps we do to yourself -- it is very important.

## Incidence matrix

First, we need to build an [incidence matrix](https://en.wikipedia.org/wiki/Incidence_matrix). You can think of pairs `(FirstId, SecondId)` as of edges in an undirected graph. 

The incidence matrix is a matrix of size `(maxId + 1, maxId + 1)`, where each row (column) `i` corresponds `i-th` `Id`. In this matrix we put the value `1` to the position `[i, j]`, if and only if a pair `(i, j)` or `(j, i)` is present in  a given set of pais `(FirstId, SecondId)`. All the other elements in the incidence matrix are zeros.   

**Important!** The incidence matrices are typically very very sparse (small number of non-zero values). At the same time incidence matrices are usually huge in terms of total number of elements, and it is **impossible to store them in memory in dense format**. But due to their sparsity incidence matrices **can be easily represented as sparse matrices**. If you are not familiar with sparse matrices, please see [wiki](https://en.wikipedia.org/wiki/Sparse_matrix) and [scipy.sparse reference](https://docs.scipy.org/doc/scipy/reference/sparse.html). Please, use any of `scipy.sparse` constructors to build incidence matrix. 

For example, you can use this constructor: `scipy.sparse.coo_matrix((data, (i, j)))`. We highly recommend to learn to use different `scipy.sparse` constuctors, and matrices types, but if you feel you don't want to use them, you can always build this matrix with a simple `for` loop. You will need first to create a matrix using `scipy.sparse.coo_matrix((M, N), [dtype])` with an appropriate shape `(M, N)` and then iterate through `(FirstId, SecondId)` pairs and fill corresponding elements in matrix with ones. 

**Note**, that the matrix should be symmetric and consist only of zeros and ones. It is a way to check yourself.

In [7]:
pairs_list = list(set(list(zip(test.SecondId,test.FirstId)) + list(zip(test.FirstId,test.SecondId)))) 
#print(pairs_list)
rows, columns = [id1 for id1, id2 in pairs_list], [id2 for id1, id2 in pairs_list]
#print(rows)
#print(columns)
ones = np.ones(len(rows)).astype(int)
inc_mat = scipy.sparse.coo_matrix((ones, (rows, columns))) # YOUR CODE GOES HERE (but probably you will need to write few more lines before)
#print(inc_mat)

# Sanity checks
assert inc_mat.max() == 1
assert inc_mat.sum() == 736872

It is convenient to have matrix in `csr` format eventually.

In [8]:
inc_mat = inc_mat.tocsr()

## Now build the magic feature

Why did we build the incidence matrix? We can think of the rows in this matix as of representations for the objects. `i-th` row is a representation for an object with `Id = i`. Then, to measure similarity between two objects we can measure similarity between their representations. And we will see, that such representations are very good.

Now select the rows from the incidence matrix, that correspond to `test.FirstId`'s, and `test.SecondId`'s.

In [9]:
# Note, scipy goes crazy if a matrix is indexed with pandas' series. 
# So do not forget to convert `pd.series` to `np.array`
# These lines should normally run very quickly 

rows_FirstId   = inc_mat[test.FirstId.values] # YOUR CODE GOES HERE
rows_SecondId  = inc_mat[test.SecondId.values] # YOUR CODE GOES HERE
print(rows_FirstId," ",rows_FirstId.shape)
print(rows_SecondId," ",rows_SecondId.shape)

  (0, 24669)	1
  (0, 24420)	1
  (0, 22945)	1
  (0, 22689)	1
  (0, 22288)	1
  (0, 20813)	1
  (0, 20558)	1
  (0, 19083)	1
  (0, 17746)	1
  (0, 17491)	1
  (0, 17301)	1
  (0, 15819)	1
  (0, 14546)	1
  (0, 14291)	1
  (0, 13829)	1
  (0, 12360)	1
  (0, 11488)	1
  (0, 11235)	1
  (0, 9366)	1
  (0, 9136)	1
  (0, 8053)	1
  (0, 7797)	1
  (0, 5648)	1
  (0, 5394)	1
  (0, 4993)	1
  :	:
  (368549, 15599)	1
  (368549, 14402)	1
  (368549, 14326)	1
  (368549, 12216)	1
  (368549, 12140)	1
  (368549, 11874)	1
  (368549, 11346)	1
  (368549, 11270)	1
  (368549, 9499)	1
  (368549, 9247)	1
  (368549, 9171)	1
  (368549, 7908)	1
  (368549, 7832)	1
  (368549, 7143)	1
  (368549, 6709)	1
  (368549, 6545)	1
  (368549, 5505)	1
  (368549, 5429)	1
  (368549, 5104)	1
  (368549, 5028)	1
  (368549, 3525)	1
  (368549, 1843)	1
  (368549, 1282)	1
  (368549, 1206)	1
  (368549, 337)	1   (368550, 26325)
  (0, 24669)	1
  (0, 24314)	1
  (0, 22945)	1
  (0, 22582)	1
  (0, 22183)	1
  (0, 20813)	1
  (0, 20453)	1
  (0, 19083)	1
  (0, 

Our magic feature will be the *dot product* between representations of a pair of objects. Dot product can be regarded as similarity measure -- for our non-negative representations the dot product is close to 0 when the representations are different, and is huge, when representations are similar. 

Now compute dot product between corresponding rows in `rows_FirstId` and `rows_SecondId` matrices.

In [10]:
# Note, that in order to do pointwise multiplication in scipy.sparse you need to use function `multiply`
# regular `*` corresponds to matrix-matrix multiplication

f = np.array(rows_FirstId.multiply(rows_SecondId).sum(axis=1)).squeeze() # YOUR CODE GOES HERE
print(f)

# Sanity check
assert f.shape == (368550, )

[20 14 20 ..., 14 14 14]


That is it! **We've built our magic feature.** 

# From magic feature to binary predictions

But how do we convert this feature into binary predictions? We do not have a train set to learn a model, but we have a piece of information about test set: the baseline accuracy score that you got, when submitting constant. And we also have a very strong considerations about the data generative process, so probably we will be fine even without a training set. 

We may try to choose a thresold, and set the predictions to 1, if the feature value `f` is higer than the threshold, and 0 otherwise. What threshold would you choose? 

How do we find a right threshold? Let's first examine this feature: print frequencies (or counts) of each value in the feature `f`.

In [11]:
# For example use `np.unique` function, check for flags

print(np.unique(f, return_counts=True)) # YOUR CODE GOES HERE

(array([14, 15, 19, 20, 21, 28, 35]), array([183279,    852,    546, 183799,      6,     54,     14]))


Do you see how this feature clusters the pairs? Maybe you can guess a good threshold by looking at the values? 

In fact, in other situations it can be not that obvious, but in general to pick a threshold you only need to remember the score of your baseline submission and use this information. Do you understand why and how?  

Choose a threshold below: 

In [12]:
pred = f > 19 # SET THRESHOLD HERE

# Finally, let's create a submission

In [ ]:
submission = test.loc[:,['pairId']]
submission['Prediction'] = pred.astype(int)

submission.to_csv('submission.csv', index=False)

Now submit it to the grader! It is not possible to submit directly from this notebook, as we need to submit a `csv` file, not a single number (limitation of Coursera platform). 

To download `submission.csv` file that you've just produced <a href='./submission.csv'>click here</a> (if the link opens in browser, right-click on it and shoose "Save link as"). Then go to [assignment page](https://www.coursera.org/learn/competitive-data-science/programming/KsASv/data-leakages/submission) and submit your `.csv` file in 'My submission' tab.


If you did everything right, the score should be very high.

**Finally:** try to explain to yourself, why the whole thing worked out. In fact, there is no magic in this feature, and the idea to use rows in the incidence matrix can be intuitively justified.

# Bonus

Interestingly, it is not the only leak in this dataset. There is another totally different way to get almost 100% accuracy. Try to find it!